edx_proctoring/api.py

```diff
@@ -0,0 +1,558 @@
+
+        raise StudentExamAttemptDoesNotExistsException(err_msg)
+
+    return _start_exam_attempt(existing_attempt)
+
+
+def start_exam_attempt_by_code(attempt_code):
+    """
+    Signals the beginning of an exam attempt when we only have
+    an attempt code
+    """
+
+    existing_attempt = ProctoredExamStudentAttempt.objects.get_exam_attempt_by_code(attempt_code)
+
+    if not existing_attempt:
+        err_msg = (
+            'Cannot start exam attempt for attempt_code = {attempt_code} '
+            'because it does not exist!'
+        ).format(attempt_code=attempt_code)
+
+        raise StudentExamAttemptDoesNotExistsException(err_msg)
+
+    return _start_exam_attempt(existing_attempt)
+
+
+def _start_exam_attempt(existing_attempt):
+    """
+    Helper method
+    """
+
+    if existing_attempt.started_at:
+        # cannot restart an attempt
+        err_msg = (
+            'Cannot start exam attempt for exam_id = {exam_id} '
+            'and user_id = {user_id} because it has already started!'
+        ).format(exam_id=existing_attempt.proctored_exam.id, user_id=existing_attempt.user_id)
+
+        raise StudentExamAttemptedAlreadyStarted(err_msg)
+
+    existing_attempt.start_exam_attempt()
+
+    return existing_attempt.id
+
+
+def stop_exam_attempt(exam_id, user_id):
+    """
+    Marks the exam attempt as completed (sets the completed_at field and updates the record)
+    """
+    exam_attempt_obj = ProctoredExamStudentAttempt.objects.get_exam_attempt(exam_id, user_id)
+    if exam_attempt_obj is None:
+        raise StudentExamAttemptDoesNotExistsException('Error. Trying to stop an exam that does not exist.')
+    else:
+        exam_attempt_obj.completed_at = datetime.now(pytz.UTC)
+        exam_attempt_obj.status = ProctoredExamStudentAttemptStatus.completed
+        exam_attempt_obj.save()
+        return exam_attempt_obj.id
+
+
+def mark_exam_attempt_timeout(exam_id, user_id):
+    """
+    Marks the exam attempt as timed_out
+    """
+    exam_attempt_obj = ProctoredExamStudentAttempt.objects.get_exam_attempt(exam_id, user_id)
+    if exam_attempt_obj is None:
+        raise StudentExamAttemptDoesNotExistsException('Error. Trying to time out an exam that does not exist.')
+    else:
+        exam_attempt_obj.status = ProctoredExamStudentAttemptStatus.timed_out
+        exam_attempt_obj.save()
+        return exam_attempt_obj.id
+
+
+def mark_exam_attempt_as_ready(exam_id, user_id):
+    """
+    Marks the exam attemp as ready to start
+    """
+
+    exam_attempt_obj = ProctoredExamStudentAttempt.objects.get_exam_attempt(exam_id, user_id)
+    if exam_attempt_obj is None:
+        raise StudentExamAttemptDoesNotExistsException('Error. Trying to time out an exam that does not exist.')
+    else:
+        exam_attempt_obj.status = ProctoredExamStudentAttemptStatus.ready_to_start
+        exam_attempt_obj.save()
+        return exam_attempt_obj.id
+
+
+def remove_exam_attempt_by_id(attempt_id):
```

Comment:
I'm not seeing any other way to remove an exam attempt, so I'm thinking we don't need to be so specific with the method signature here -- I think remove_exam_attempt(attempt_id) should suffice.